packages/cosmic-swingset/src/fees.js

```diff
@@ -0,0 +1,133 @@
+import { assert, details as X } from '@agoric/assert';
+
+import { DecimalMath } from './rational.js';
+
+// This controls how precise the fee calculation is.
+export const DEFAULT_COUNT_DECIMAL_PLACES = 12n;
+
+// Logic adapted from @agoric/swingset-vat/src/vats/comms/delivery.js
+const countSlots = message => {
```

Comment:
This makes me nervous, and it crosses so many abstractions boundaries at once. The comms protocol (the formatter and parser) should be owned by comms, and having a fragile copy of its parser live all the way out here feels bad.

What if we added a new ("deep") export from SwingSet, like `commsCountSlots(message) -> int`, to encapsulate this function? And the source for that function would live in `SwingSet/src/vats/comms/`. It'd still be a serious layering violation, but at least knowledge of the format could be confined to a single directory.

I'm on the fence as to whether including num(slots) in the fee calculation is worth the layering violation. It's true that a lot of the overhead of dealing with a message will be copying slot strings around (or translating them, etc), so including that count in the fee equation will make it more accurate. But it doesn't yet capture the potential for new imports or exports (which incur some sort of long-term obligation and cost), and I can't decide whether those are likely to overwhelm whatever marginal cost we measure for just the presence of a slot at all.

Reply:
Withdrawn.